pilot/pkg/config/coredatamodel/discovery.go

```diff
@@ -53,34 +51,32 @@ func (o *DiscoveryOptions) mixerEnabled() bool {
 	return o.Env != nil && o.Env.Mesh != nil && (o.Env.Mesh.MixerCheckServer != "" || o.Env.Mesh.MixerReportServer != "")
 }
 
-// MCPDiscovery provides storage for NotReadyEndpoints
+// MCPDiscovery provides discovery interface for SyntheticServiceEntries
 type MCPDiscovery struct {
-	notReadyEndpointsMu sync.RWMutex
-	// [ip:port]config
-	notReadyEndpoints map[string]*model.Config
+	*SyntheticServiceEntryController
 	*DiscoveryOptions
 }
 
 // NewMCPDiscovery provides a new instance of Discovery
-func NewMCPDiscovery(options *DiscoveryOptions) *MCPDiscovery {
+func NewMCPDiscovery(controller CoreDataModel, options *DiscoveryOptions) *MCPDiscovery {
 	return &MCPDiscovery{
-		notReadyEndpoints: make(map[string]*model.Config),
-		DiscoveryOptions:  options,
+		SyntheticServiceEntryController: controller.(*SyntheticServiceEntryController),
+		DiscoveryOptions:                options,
 	}
 }
 
-// RegisterNotReadyEndpoints registers newly received NotReadyEndpoints
-// via MCP annotations
-func (d *MCPDiscovery) RegisterNotReadyEndpoints(conf *model.Config) {
-	d.notReadyEndpointsMu.Lock()
-	defer d.notReadyEndpointsMu.Unlock()
+// Services list declarations of all SyntheticServiceEntries in the system
+func (d *MCPDiscovery) Services() ([]*model.Service, error) {
+	services := make([]*model.Service, 0)
 
-	if nrEps, ok := conf.Annotations[notReadyEndpointkey]; ok {
-		addrs := strings.Split(nrEps, ",")
-		for _, addr := range addrs {
-			d.notReadyEndpoints[addr] = conf
-		}
+	syntheticServiceEntries, err := d.List(schemas.SyntheticServiceEntry.Type, model.NamespaceAll)
+	if err != nil {
+		return nil, err
+	}
+	for _, cfg := range syntheticServiceEntries {
+		services = append(services, convertServices(cfg)...)
```

Comment:
As a future optimization, we should think about doing this conversion once.

Reply:
+1 I created an issue which only focuses on the optimization in the discovery.

Reply:
cache the services